docs/architecture/adr-038-state-listening.md

```diff
@@ -182,349 +137,587 @@ func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
-			}()
-		} else {
-			// fire and forget semantics
-			go func() {
-				if err := streamingListener.ListenEndBlock(app.deliverState.ctx, req, res); err != nil {
-					app.logger.Error("EndBlock listening hook failed", "height", req.Height, "err", err)
-				}
-			}()
-		}
-	}
-	// wait for all the listener calls to finish
-	wg.Wait()
+    var abciRes abci.ResponseDeliverTx
+    defer func() {
+        // call the streaming service hook with the EndBlock messages
+        if app.abciListener != nil {
+            ctx := app.deliverState.ctx
+            blockHeight := ctx.BlockHeight()
+            if app.stopNodeOnStreamingErr {
+                if err := app.abciListener.ListenDeliverTx(blockHeight, req, res); err != nil {
+                    app.logger.Error("DeliverTx listening hook failed", "height", blockHeight, "err", err)
+                    os.Exit(1)
+                }
+            } else {
+                go func(req abci.RequestDeliverTx, res abci.ResponseDeliverTx) {
+                    if reqErr == nil && resErr == nil {
+                        if err := app.abciListener.ListenDeliverTx(blockHeight, req, res); err != nil {
+                            app.logger.Error("DeliverTx listening hook failed", "height", blockHeight, "err", err)
+                        }
+                    }
+                }(req, abciRes)
+            }
+        }
+    }()
 
-	return res
+    ...
+
+    return abciRes
 }
 ```
 
 ```go
-func (app *BaseApp) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
-	
-	var abciRes abci.ResponseDeliverTx
-	defer func() {
-		// call the hooks with the BeginBlock messages
-		wg := new(sync.WaitGroup)
-		for _, streamingListener := range app.abciListeners {
-			streamingListener := streamingListener // https://go.dev/doc/faq#closures_and_goroutines
-			if streamingListener.HaltAppOnDeliveryError() {
-				// increment the wait group counter
-				wg.Add(1)
-				go func() {
-					// decrement the counter when the go routine completes
-					defer wg.Done()
-					if err := streamingListener.ListenDeliverTx(app.deliverState.ctx, req, abciRes); err != nil {
-						app.logger.Error("DeliverTx listening hook failed", "err", err)
-						app.halt()
-					}
-				}()
-			} else {
-				// fire and forget semantics
-				go func() {
-					if err := streamingListener.ListenDeliverTx(app.deliverState.ctx, req, abciRes); err != nil {
-						app.logger.Error("DeliverTx listening hook failed", "err", err)
-					}
-				}()
-			}
-		}
-		// wait for all the listener calls to finish
-		wg.Wait()
-	}()
-	
-	...
+func (app *BaseApp) Commit() abci.ResponseCommit {
 
-	return res
+    ...
+
+    res := abci.ResponseCommit{
+        Data:         commitID.Hash,
+        RetainHeight: retainHeight,
+    }
+
+    // call the streaming service hook with the Commit messages
+    if app.abciListener != nil {
+        ctx := app.deliverState.ctx
+        blockHeight := ctx.BlockHeight()
+        changeSet := app.cms.PopStateCache()
```

Comment:
The idea of pass change set as a whole is to avoid some allocations for multiple streaming services, but considering different stream services can monitor different subset of store keys, we need to filter the change set based on the keys, I guess it'll defeat the design purpose.

Reply:
For simplicity's sake, I would argue the key set filtering should exist globally, not on a per plugin/consumer based level.

Reply:
I guess for the out-of-process plugins, if we don't filter change set before wire transfer, we'll need to transfer the big set multiple times, even if different plugins only care about different stores.

Reply:
One benefit that we get with plugins over gRPC is the isolation of state streaming to a single plugin (vs the old design). Having multiple plugins filtering out state change events would bare the unnecessary cost of creating additional gRPC services just to filter out before wire transfer. There is no cost saving here. It puts additional burden on the SDK. You're also exposed to plugins overlapping on the store keys they listen. This is why I changed `streaming.plugins = []` to `streaming.plugin=""`. Do the filtering (and fan-out) downstream.

Different plugins should be reserved for exposing different parts of the system and not over the same data.